plugins/scale/scale_title_filter.cpp

```diff
@@ -0,0 +1,322 @@
+            return title.find(filter) != std::string::npos;
+        }
+
+        return view->get_title().find(title_filter) != std::string::npos;
+    }
+
+  public:
+    void init() override
+    {
+        grab_interface->name = "scale_title_filter";
+        grab_interface->capabilities = 0;
+
+        output->connect_signal("scale-filter", &view_filter);
+        output->connect_signal("scale-end", &scale_end);
+    }
+
+    void fini() override
+    {
+        clear_overlay();
+        output->disconnect_signal(&view_filter);
+        wf::get_core().disconnect_signal(&scale_key);
+        output->disconnect_signal(&scale_end);
+    }
+
+    wf::signal_connection_t view_filter{[this] (wf::signal_data_t *data)
+        {
+            if (!scale_running)
+            {
+                wf::get_core().connect_signal("keyboard_key", &scale_key);
+                scale_running = true;
+            }
+
+            auto v   = static_cast<scale_filter_signal*>(data);
+            v->hide |= !should_show_view(v->view);
+        }
+    };
+
+    wf::signal_connection_t scale_key{[this] (wf::signal_data_t *data)
+        {
+            auto k =
+                static_cast<wf::input_event_signal<wlr_event_keyboard_key>*>(data);
+            if ((k->event->state != WL_KEYBOARD_KEY_STATE_PRESSED) ||
+                (k->event->keycode == KEY_ESC) || (k->event->keycode == KEY_ENTER))
+            {
+                return;
+            }
+
+            if (k->event->keycode == KEY_BACKSPACE)
+            {
+                if (!title_filter.empty())
+                {
+                    int len = char_len.back();
+                    char_len.pop_back();
+                    for (int i = 0; i < len; i++)
+                    {
+                        title_filter.pop_back();
+                    }
+                } else
+                {
+                    return;
+                }
+            } else
+            {
+                auto seat     = wf::get_core().get_current_seat();
+                auto keyboard = wlr_seat_get_keyboard(seat);
+                if (!keyboard)
+                {
+                    return; /* should not happen */
+                }
+
+                auto xkb_state = keyboard->xkb_state;
+
+                /* taken from libxkbcommon guide */
+                int keycode = k->event->keycode + 8;
+                int size    = xkb_state_key_get_utf8(xkb_state, keycode, nullptr, 0);
+                if (size <= 0)
+                {
+                    return;
+                }
+
+                std::string tmp(size, 0);
+                xkb_state_key_get_utf8(xkb_state, keycode, tmp.data(), size + 1);
+                char_len.push_back(size);
+                title_filter += tmp;
+            }
+
+            LOGI("Title filter changed: ", title_filter);
+            scale_activate_signal signal;
+            output->emit_signal("scale-activate", &signal);
```

Comment:
It is not needed by the title filter, but it would be needed by another plugin that wants to start scale, e.g. the simple test plugin I had included. That can be adapted to use the toggle bindings though.